pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go

```diff
@@ -254,31 +269,85 @@ var planMutators = []mutator{
-		for _, s := range p.services() {
-			s.startUpgrade(upgrade.to)
+	upgradeStepsForService := func(
+		service *ServiceContext, from, to *clusterupgrade.Version, virtualClusterSetup, scheduleHooks bool,
+	) []testStep {
+		if p.deploymentMode == SeparateProcessDeployment {
+			// Separate-process deployments are different because the two
+			// services involved (system and tenant) restart at different
+			// times (since they run separate binaries): the system (storage
+			// cluster) needs to upgrade to completion first, and only then
+			// the tenant is allowed to start upgrading.
+			//
+			// To allow for internal and external hooks to recognize this
+			// stage, we set the service's stage accordingly below.
+			if service.IsSystem() {
+				// Indicates that tenant is waiting for system to finish
+				// upgrading.
+				p.currentContext.Tenant.Stage = UpgradingSystemStage
+			} else {
+				// Indicates that system is waiting for tenant to finish
+				// upgrading.
+				p.currentContext.System.Stage = UpgradingTenantStage
+			}
+		}
+
+		var steps []testStep
+		addSteps := func(ss []testStep) {
+			steps = append(steps, ss...)
 		}
-		upgrade.Add(p.initUpgradeSteps(virtualClusterSetup))
-		if p.shouldRollback(upgrade.to) {
+
+		addSteps(p.initUpgradeSteps(service, virtualClusterSetup))
+		if p.shouldRollback(to) {
 			// previous -> next
-			upgrade.Add(p.upgradeSteps(
-				TemporaryUpgradeStage, upgrade.from, upgrade.to, scheduleHooks, virtualClusterSetup,
+			addSteps(p.upgradeSteps(
+				service, TemporaryUpgradeStage, from, to, scheduleHooks, virtualClusterSetup,
 			))
 			// next -> previous (rollback)
-			upgrade.Add(p.downgradeSteps(
-				upgrade.to, upgrade.from, scheduleHooks, virtualClusterSetup,
+			addSteps(p.downgradeSteps(
+				service, to, from, scheduleHooks, virtualClusterSetup,
 			))
 		}
 		// previous -> next
-		upgrade.Add(p.upgradeSteps(
-			LastUpgradeStage, upgrade.from, upgrade.to, scheduleHooks, virtualClusterSetup,
+		addSteps(p.upgradeSteps(
+			service, LastUpgradeStage, from, to, scheduleHooks, virtualClusterSetup,
 		))
 
 		// finalize -- i.e., run upgrade migrations.
-		upgrade.Add(p.finalizeUpgradeSteps(upgrade.from, upgrade.to, scheduleHooks, virtualClusterSetup))
+		addSteps(p.finalizeUpgradeSteps(service, from, to, scheduleHooks, virtualClusterSetup))
 
 		// run after upgrade steps, if any,
-		upgrade.Add(p.afterUpgradeSteps(upgrade.from, upgrade.to, scheduleHooks))
+		addSteps(p.afterUpgradeSteps(service, from, to, scheduleHooks))
+
+		return steps
+	}
+
+	planUpgrade := func(upgrade *upgradePlan, virtualClusterSetup, scheduleHooks bool) {
+		for _, s := range p.services() {
+			s.startUpgrade(upgrade.to)
+		}
+
+		systemUpgradeSteps := upgradeStepsForService(
+			p.currentContext.System, upgrade.from, upgrade.to, virtualClusterSetup, scheduleHooks,
+		)
+
+		if p.deploymentMode != SeparateProcessDeployment || !virtualClusterSetup {
+			upgrade.Add(systemUpgradeSteps)
```

Comment:
I had to pause here to remember that `systemUpgradeSteps` is essentially equivalent to non-multitenant upgrades; it might be worth dropping an explicit comment as a reminder.

Reply:
Comment added.